working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+`dynamic`. The forwarding expression `this.floor()` in the implicitly
+induced getter will invoke the instance method, which is subject to late
+binding (so we may end up running `int.floor()` or `double.floor()`,
+depending on the dynamic type of `this`).*
+
+At run time, for a given instance `o` typed as an extension type `E`, there
+is _no_ reification of `E` associated with `o`.
+
+*This means that, at run time, an object never "knows" that it is being
+viewed as having an extension type. By soundness, the run-time type of `o`
+will be a subtype of the on-type of `E`.*
+
+The run-time representation of a type argument which is a non-protected
+extension type `E` (respectively 
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>) 
+is the corresponding instantiated on-type.
+
+*This means that a non-protected extension type and the underlying on-type
+are considered as being the same type at run time. So we can freely use a
+cast to introduce or discard the extension type, as the static type of an
+instance, or as a type argument in the static type of a data structure or
+function involving the extension type.*
+
+The run-time representation of a type argument which is a protected
+extension type `E` (respectively 
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>) 
+is an identification of `E` (respectively 
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>).
+
+*In particular, it is not the same as the run-time representation of the
+corresponding on-type. This is necessary in order to maintain that the
+on-type and the protected extension type are unrelated.*
+
+*For a protected extension type `E`, with a data structure or function where
+`E` occurs as a subterm in the type (that is, as a type argument, a return
+type, or a parameter type), a cast that tries to introduce or discard the
+protected extension type will fail at run time.*
+
+*In the non-protected case this treatment may appear to be
+unsound. However, it is in fact sound: Let `E` be a non-protected
+extension type with on-type `T`. This implies that `void Function(E)` is
+represented as `void Function(T)` at run-time. In other words, it is
+possible to have a variable of type `void Function(T)` that refers to a
+function object of type `void Function(E)`. This seems to be a soundness
+violation because `T <: E` and not vice versa, statically. However, we
+consider such types to be the same type at run time, which is in any case
+the finest distinction that we can maintain because there is no
+representation of `E` at run time. There is no soundness issue, because
+the added discipline of a non-protected extension type is voluntary.*
+
+A type test, `o is U`, and a type cast, `o as U`, where `U` is or contains
+an extension type, is performed at run time as a type test and type cast on
+the run-time representation of the extension type as described above.
+
+*Note that `U` cannot be a protected extension type, because the expression
```

Comment:
It can be dynamically. 

```dart
bool foo<T>(Object x) => x is T;
protected extension type E on int {}
void main() {
  foo<E>(3);
}
```

Reply:
The cast to `U` is then a dynamic error (or it would call `verifyThis`, if we have that).

Reply:
As discussed elsewhere, I don't know how to make this work out.